csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+//
+// The concept is that the following program:
+//   tv1 = UnaryOp(tv0)
+//   tv2 = PadOp(tv1)
+// can be replayed as the program below:
+//   tv0_padded = PadOp(tv0)
+//   tv2_new = UnaryOp(tv0_padded)
+// Given certain constraints on the UnaryOp, we can ensure that the
+// computational output remains the same when we replace all uses of `tv2` with
+// `tv2_new`.
+//
+// This function only propagates simple padding, where its pad value is `zero`
+// (or `false` for boolean) and pad widths are non-negative. This allows us to
+// unconditionally merge neighboring PadOps as a single op. We also restrict
+// propagation on operatoins that can allow PadOp to propagated across. See
+// `zeroIsFixedPoint` and `zeroIsIdentity`.
+void propagatePad(Fusion* fusion) {
+  // propagating PadOp
+  auto exprs = fusion->exprs();
+  auto filtered_pads = ir_utils::filterByType<PadOp>(exprs);
+  std::vector<PadOp*> frontier;
+  frontier.reserve(filtered_pads.size());
+
+  // NOTE: we only consider simple padop as propagation frontier.
+  std::copy_if(
+      filtered_pads.begin(),
+      filtered_pads.end(),
+      std::back_inserter(frontier),
+      isSimplePadOp);
+
+  // NOTE: this is a WAR. We use a set of `simple_pad_set` to track all mergable
+  // PadOps, this is to leverage the assumption that all PadOps before CatOps
+  // are simple op, but might not be able to be evaluated as such during compile
+  // time.
+  std::unordered_set<PadOp*> simple_pad_set(frontier.begin(), frontier.end());
+
+  while (!frontier.empty()) {
+    PadOp* p = frontier.back();
+    frontier.pop_back();
+
+    // if no uses, this has already been short-wired.
+    if (p->out()->uses().empty() && !p->out()->isFusionOutput()) {
+      continue;
+    }
+
+    // unify all consumer pad of tv;
+    TensorView* tv = p->in()->as<TensorView>();
+    for (Expr* use : tv->uses()) {
+      if (use == p) {
+        continue;
+      }
+      // check if use is the same pad operation (same pad value / width e.t.c.)
+      if (isSamePadOp(use, p)) {
+        // replace consumer of use->out() with p->out()
+        ir_utils::replaceValInAllExprInputsAndFusionOutputs(
+            use->output(0), p->out());
+        fusion->removeExpr(use);
+      }
+    }
+
+    // if tv is fusion output, we need to keep tv alive, it might render
+    // propagating PadOp before tv->definition() being non-optimal.
+    if (tv->isFusionOutput()) {
+      continue;
+    }
+
+    // check for pad_dependencies to verify that 'p' can be moved before 'def'.
+    std::unordered_set<Val*> pad_inputs;
+    for (Val* val : p->inputs()) {
+      if (val == p->in() || val->isConst()) {
+        continue;
+      }
+      pad_inputs.insert(val);
+    }
+    std::unordered_set<Val*> pad_dependencies =
+        DependencyCheck::getAllDependentVals(pad_inputs);
+    auto pad_replay_check = [&pad_dependencies](Expr* expr) {
+      return std::all_of(
+          expr->inputs().begin(),
+          expr->inputs().end(),
```

Comment:
I don't follow this part. IIUC, `pad_replay_check` tries to detect whether swapping `p` and the preceding `expr` will create cycles. However, having both `expr` and `p` use the same `Val*` doesn't necessarily create cycles. `expr->outputs` would be more reasonable to me.

Reply:
:bow: 

good catch! I messed it up in the refactor. Thanks a lot!
I'll make sure that I create a test for this one.